File: music/data/raw/spotify/spotify-models/src/main/java/yurykorzun/art/universe/music/data/raw/spotify/domain/entity/SpotifyTrack.java
package yurykorzun.art.universe.music.data.raw.spotify.domain.entity;

import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.SuperBuilder;
import yurykorzun.art.universe.music.data.raw.spotify.common.SpotifyConstants;
import yurykorzun.art.universe.music.data.raw.spotify.domain.entity.common.BaseSpotifyEntity;
import yurykorzun.art.universe.music.data.raw.spotify.enums.SpotifyEntityType;

@Entity(name = "track")
@SuperBuilder
@NoArgsConstructor
@Getter @Setter
public class SpotifyTrack extends BaseSpotifyEntity {

    @Id
    @SequenceGenerator(name = "track_seq_gen", sequenceName = "track_seq", allocationSize = SpotifyConstants.JDBC_ENTITY_SEQ_ALLOCATION)
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "track_seq_gen")
    @Setter(AccessLevel.NONE)
    private long id;

    @Column(name = "duration_ms")
    private Integer durationMs;

    @Column(name = "track_number")
    private Integer trackNumber;

    @Column(name = "disc_number")
    private Integer discNumber;

    @Column(name = "has_explicit_lyrics")
    private Boolean hasExplicitLyrics;

    @Column(name = "is_playable")
    private Boolean isPlayable;

    @Column(name = "spotify_url")
    private String spotifyUrl;

    @Column(name = "uri")
    private String uri;

    @Column(name = "isrc")
    private String isrc;

    @Column(name = "ean")
    private String ean;

    @Column(name = "upc")
    private String upc;

    @Column(name = "primary_artist_id")
    private Long primaryArtistId;

    @Column(name = "primary_artist_spotify_id")
    private String primaryArtistSpotifyId;

    @Column(name = "album_id")
    private Long albumId;

    @Column(name = "album_spotify_id")
    private String albumSpotifyId;

    @Override
    public SpotifyEntityType getEntityType() {
        return SpotifyEntityType.TRACK;
    }
}

File: ui/src/music/data/raw/spotify/types/spotify-track.ts
import { BaseSpotifyEntity } from "./spotify-base-entity.ts";
import { type ArtistRelatedRawEntity } from "@/music/shared/types/entities.ts";
import type { Track } from "@/music/shared/types/entities.ts";

export class SpotifyTrack
    extends BaseSpotifyEntity<"track">
    implements ArtistRelatedRawEntity<"track">
{
    spotifyId: string;
    spotifyUrl: string | null;
    durationMs: number | null;
    trackNumber: number | null;
    primaryArtistId: number | null;
    primaryArtistName: string | null;
    albumId: number | null;

    constructor(
        id: number,
        name: string,
        spotifyId: string,
        spotifyUrl: string | null,
        durationMs: number | null,
        trackNumber: number | null,
        primaryArtistId: number | null,
        primaryArtistName: string | null,
        albumId: number | null,
        masterEntity?: Track
    ) {
        super(id, name, masterEntity);
        this.spotifyId = spotifyId;
        this.spotifyUrl = spotifyUrl;
        this.durationMs = durationMs;
        this.trackNumber = trackNumber;
        this.primaryArtistId = primaryArtistId;
        this.primaryArtistName = primaryArtistName;
        this.albumId = albumId;
    }

    getEntityType(): "track" {
        return "track";
    }

    get artist(): { id: number; name: string } | undefined {
        if (this.primaryArtistId && this.primaryArtistName) {
            return { id: this.primaryArtistId, name: this.primaryArtistName };
        }
        return undefined;
    }

    getExternalArtistId(): number | undefined {
        return this.primaryArtistId ?? undefined;
    }

    getMasterArtistId(): number | undefined {
        return this.getMasterEntity()?.primaryArtistId;
    }
}
